### PR TITLE
Prevent AnonymousKernelCommands from ending up inside serialized KernelEvents

### PR DESCRIPTION
**Pull request by shyamnamboodiripad**
AnonymousKernelCommands are not supposed to be serialized across process boundaries and should therefore not be included as part of KernelEvents.

### CODE REVIEW

File: src/Microsoft.DotNet.Interactive/Parsing/SubmissionParser.cs
Comment (shyamnamboodiripad) [inline src/Microsoft.DotNet.Interactive/Parsing/SubmissionParser.cs:127]: There were some inadvertent whitespace fix ups - but this is the main part of the fix.

File: src/Microsoft.DotNet.Interactive/Events/KernelEvent.cs
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/Events/KernelEvent.cs:19]: I don't think this is valid. This API isn't opinionated about serialization concerns and there are valid use cases for non-serializable command types.

Comment (shyamnamboodiripad) [reply src/Microsoft.DotNet.Interactive/Events/KernelEvent.cs:19]: Ah it is legal to have non-serializable commands in events? My assumption was that the system isn't opinionated about whether events flow in-process or out-of-process and that therefore non-serializable commands should be avoided in all events (to avoid cases where an event works fine when the target kernel is in-process - but fails when the same kernel is moved out-of-process).

Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive/Events/KernelEvent.cs:19]: The design opinion is that out-of-process communication isn't required or assumed.

File: src/Microsoft.DotNet.Interactive.Tests/Parsing/SubmissionParserTests.cs
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Tests/Parsing/SubmissionParserTests.cs:370]: This is intended to be generally true, so not specific to `DiagnosticsProduced`. The internal command (`AnonymousKernelCommand` in this case, but it could as easily be one of the public command types) is an implementation detail that the external caller shouldn't have to be aware of.

It might be useful to add another similar test for the split between subkernels because the same should be true for, for example:

```csharp
#!csharp
123
#!fsharp
456
```

Comment (shyamnamboodiripad) [reply src/Microsoft.DotNet.Interactive.Tests/Parsing/SubmissionParserTests.cs:370]: So looks like this invariant is not satisfied for all events in the above case. For example, `CodeSubmissionReceived`, `CompleteCodeSubmissionReceived`, `ReturnValueProduced` events for the repro code you shared above point back to the per-language `SubmitCode` commands created internally for C# and F# respectively. Only the final `CommandSucceded` event points back to the original `SubmitCode` command.

If I change the submission to the following then interestingly, the `DiagnosticsProduced` event for the erroneous line points to the F#-specific `SubmitCode` command with code `some error`. However, the line spans in the `DiagnosticsProduced.Diagnostics` reflect the spans in the original submission.
```
#!csharp
123
 #!fsharp
some error
```

This may be *reasonable* considering that the inner commands reflect the expected command splitting behavior - however, it also seems like internal implementation details of the callee are bleeding through to the caller.

If this is unexpected, I can log a separate bug to a) fix the behavior and b) add tests to ensure that inner commands are never exposed to the caller. Thoughts?

File: src/Microsoft.DotNet.Interactive/Events/KernelEvent.cs
Comment (shyamnamboodiripad) [reply src/Microsoft.DotNet.Interactive/Events/KernelEvent.cs]: Got it. I've reverted the change and also updated #2884 to capture what we discussed offline (i.e. when an internal / non-serializable command is encountered, perhaps we should only serialize the properties present on the base `KernelCommand` type + add better tests to catch cases where we inadvertently end up including internal commands in `KernelEvent` payloads).

File: src/Microsoft.DotNet.Interactive.Tests/Parsing/SubmissionParserTests.cs
Comment (shyamnamboodiripad) [reply src/Microsoft.DotNet.Interactive.Tests/Parsing/SubmissionParserTests.cs:370]: @jonsequitur Btw the current PR should be ready to merge. The above issue needs more investigation / follow-up and I'd prefer not to hold up the current PR for that. Please approve if you agree :) Thanks!

File: src/Microsoft.DotNet.Interactive/Connection/ProxyKernel.cs
Comment (colombod) [inline src/Microsoft.DotNet.Interactive/Connection/ProxyKernel.cs:163]: this could lead to a deadlock if the anonymous command targets the proxy. For sure this will throw exception as it will try to forward the command on the sender and will fail at serialization. This looks like a test gap to pinpoint

**Review by colombod — CHANGES_REQUESTED** [review]
anonymous command should not leave the current host and should be handled locally even by proxy kernel

File: src/Microsoft.DotNet.Interactive/Connection/ProxyKernel.cs
Comment (shyamnamboodiripad) [reply src/Microsoft.DotNet.Interactive/Connection/ProxyKernel.cs:163]: Yes this PR is fixing one specific case where we were serializing AnonymousKernelCommand (inside SubmissionParser.cs). But the larger test gap has not been addressed yet - I have included this in #2884 and plan to address that in a follow up PR.

**Comment by shyamnamboodiripad** [comment]
> anonymous command should not leave the current host and should be handled locally even by proxy kernel

@colombod Agreed. This is part of what Jon and I discussed, and I have included the larger problem in this follow up issue - #2884 which I plan to look into soon.

This PR addresses one known case where AnonymousKernelCommand ends up leaving the current host today (see the fix in SubmissionParser.cs line 127 in the current PR). The plan that @jonsequitur and I discussed was to merge this fix now and tackle the larger issue in subsequent PR. I was essentially waiting for him to take a second look and approve the PR today.

Note that I was planning to merge this change today since this is one of the issues that came up when testing the VS side integration. So would appreciate if you could remove the block on the PR :)

**Comment by colombod** [comment]
If @jonsequitur approves this should auto merge.

**Comment by jonsequitur** [comment]
Let's find some time to discuss how we want to do this while removing rather than adding code. There's too much complexity here.

**Review by jonsequitur — APPROVED** [review]

**Review by colombod — APPROVED** [review]

**Pull request merged**
